Heatington/Optimizer/Optimizer.cs

```diff
@@ -1,123 +1,210 @@
+        {
+            return Math.Round((demand / productionCapacity), 4);
+        }
     }
 
-    private int CalculateHeatUnitsRequired(FrancescoEnergyData entry)
+    private void CalculateHeatUnitsRequired()
     {
-        double totProductionCapacity = 0;
-        int i = 0;
+        if (_dataPoints == null)
+        {
+            return;
+        }
 
-        totProductionCapacity += _pUnits[i].MaxHeat;
+        List<ResultHolder> results = new List<ResultHolder>();
 
-        while (entry.HeatDemandMwh > totProductionCapacity)
+        foreach (DataPoint dataPoint in _dataPoints)
         {
-            totProductionCapacity += _pUnits[i].MaxHeat;
-            i++;
+            // Calculates how many boilers are needed to satisfy heat demand
+            // Implies that they are ordered by efficiency
+            int nOfBoilers = SatisfyHeatDemand(dataPoint);
+
+            //Console.WriteLine("Number of boilers: {0}", nOfBoilers);
+
+            // Selects the correct boilers based on how many need to be activated
+            // Again implies that they are ordered by efficiency
+            List<ProductionUnit> selectedBoilers = SelectBoiler(nOfBoilers);
+
+            //Console.WriteLine("Boilers selected:");
+            //selectedBoilers.ForEach(Console.WriteLine);
+
+            // Sets the operation point of the selected boilers
+            selectedBoilers = SetOperationPoint(selectedBoilers, dataPoint.HeatDemand);
+
+            // Creates an object which holds the result
+            ResultHolder result = new ResultHolder(dataPoint.StartTime, dataPoint.EndTime, dataPoint.HeatDemand,
+                dataPoint.ElectricityPrice, 0, selectedBoilers);
+            // Adds the object to the list
+            results.Add(result);
+            Console.WriteLine(results[^1]);
+            Console.WriteLine(result);
+        }
 
-            if (i > _pUnits.Count)
+        Results = results;
+        results.ForEach(Console.WriteLine);
+
+
+        int SatisfyHeatDemand(DataPoint dataPoint)
+        {
+            double currentProductionCapacity = 0;
+            int i = 0;
+
+            while (dataPoint.HeatDemand > currentProductionCapacity)
             {
-                Console.WriteLine("WARNING: HEAT DEMAND CAN NOT BE SATISFIED");
-                throw new Exception("WARNING: HEAT DEMAND CAN NOT BE SATISFIED");
+                currentProductionCapacity = currentProductionCapacity + _productionUnits[i].MaxHeat;
+                i++;
+
+                if (i > _productionUnits.Count)
+                {
+                    Console.WriteLine("WARNING: HEAT DEMAND CAN NOT BE SATISFIED");
+                    throw new Exception("WARNING: HEAT DEMAND CAN NOT BE SATISFIED");
+                }
             }
+            return i;
         }
-        return i;
-    }
 
-    // Method which calculates Hourly NetProductionCost for heat only boilers
+        List<ProductionUnit> SelectBoiler(int j)
```

Comment:
Does this actually "select" the boiler?

Reply:
Yes, it selects the boiler objects from the production Unit list